lib/Parse/ParseDecl.cpp

```diff
@@ -2850,6 +2805,80 @@ ParserResult<TypeDecl> Parser::parseDeclTypeAlias(bool WantDefinition,
   return DCC.fixupParserResult(Status, TAD);
 }
 
+/// \brief Parse an associatedtype decl.
+///
+/// \verbatim
+///   decl-associatedtype:
+///     'associatedtype' identifier inheritance? ('=' type)?
+/// \endverbatim
+
+ParserResult<TypeDecl> Parser::parseDeclAssociatedType(Parser::ParseDeclOptions Flags,
+                                                       DeclAttributes &Attributes) {
+  SourceLoc AssociatedTypeLoc;
+  ParserStatus Status;
+  Identifier Id;
+  SourceLoc IdLoc;
+  
+  // Look for 'typealias' here and diagnose a fixit because parseDeclTypeAlias can
+  // ask us to fix up leftover Swift 2 code intending to be an associatedtype.
+  if (Tok.is(tok::kw_typealias)) {
+    AssociatedTypeLoc = consumeToken(tok::kw_typealias);
+    diagnose(AssociatedTypeLoc, diag::typealias_inside_protocol_without_type)
+        .fixItReplace(AssociatedTypeLoc, "associatedtype");
+    Status.setIsParseError();
```

Comment:
You don't need to set Status.setIsParseError(), since you've recovered completely via the Fix-It. That way, we can continue as if the user never made the mistake at all.